api/api.go

```diff
@@ -0,0 +1,512 @@
+package api
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+
+	"github.com/docker/libnetwork"
+	"github.com/docker/libnetwork/netutils"
+	"github.com/docker/libnetwork/sandbox"
+	"github.com/gorilla/mux"
+)
+
+var (
+	successResponse  = responseStatus{Status: "Success", StatusCode: http.StatusOK}
+	createdResponse  = responseStatus{Status: "Created", StatusCode: http.StatusCreated}
+	mismatchResponse = responseStatus{Status: "Body/URI parameter mismatch", StatusCode: http.StatusBadRequest}
+)
+
+const (
+	urlNwName = "name"
+	urlNwID   = "id"
+	urlEpName = "endpoint-name"
+	urlEpID   = "endpoint-id"
+)
+
+// NewHTTPHandler creates and initialize the HTTP handler to serve the requests for libnetwork
+func NewHTTPHandler(c libnetwork.NetworkController) func(w http.ResponseWriter, req *http.Request) {
+	h := &httpHandler{c: c}
+	h.initRouter()
+	return h.handleRequest
+}
+
+type responseStatus struct {
+	Status     string
+	StatusCode int
+}
+
+func (r *responseStatus) isOK() bool {
+	return r.StatusCode == http.StatusOK || r.StatusCode == http.StatusCreated
+}
+
+type processor func(c libnetwork.NetworkController, vars map[string]string, body []byte) (interface{}, *responseStatus)
+
+type httpHandler struct {
+	c libnetwork.NetworkController
+	r *mux.Router
+}
+
+func (h *httpHandler) handleRequest(w http.ResponseWriter, req *http.Request) {
+	// Make sure the service is there
+	if h.c == nil {
+		http.Error(w, "NetworkController is not available", http.StatusServiceUnavailable)
+		return
+	}
+
+	// Get handler from router and execute it
+	h.r.ServeHTTP(w, req)
+}
+
+func (h *httpHandler) initRouter() {
+	m := map[string]map[string]processor{
+		"GET": {
+			"/networks":                                                                   procGetNetworks,
+			"/networks/name/{" + urlNwName + ":.*}":                                       procGetNetworkByName,
+			"/networks/id/{" + urlNwID + ":.*}":                                           procGetNetworkByID,
+			"/networks/name/{" + urlNwName + ":.*}/endpoints/":                            procGetEndpoints,
+			"/networks/name/{" + urlNwName + ":.*}/endpoints/name/{" + urlEpName + ":.*}": procGetEndpointByName,
+			"/networks/name/{" + urlNwName + ":.*}/endpoints/id/{" + urlEpID + ":.*}":     procGetEndpointByID,
+		},
+		"POST": {
+			"/networks/name/{" + urlNwName + ":.*}":                                       procCreateNetwork,
+			"/networks/name/{" + urlNwName + ":.*}/endpoint/{" + urlEpName + ":.*}":       procCreateEndpoint,
+			"/networks/name/{" + urlNwName + ":.*}/endpoint/{" + urlEpName + ":.*}/join":  procJoinEndpoint,
```

Comment:
Instead of calling out Join & Leave as part of the URL Resource, I believe it is better to do `POST` on 'network-id/endipoint-id/container-id' for Join and `DELETE` on network-id/endipoint-id/container-id' for Leave.